zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs

```diff
@@ -0,0 +1,880 @@
+    },
+    gadget::{
+        evm_word::encode,
+        is_zero::{IsZeroChip, IsZeroConfig, IsZeroInstruction},
+    },
+    util::Expr,
+};
+use bus_mapping::{
+    eth_types::{ToBigEndian, Word},
+    evm::OpcodeId,
+};
+use halo2::{
+    circuit::{Layouter, Region},
+    plonk::{
+        Advice, Column, ConstraintSystem, Error, Fixed, Selector, VirtualCells,
+    },
+    poly::Rotation,
+};
+use keccak256::plain::Keccak;
+use pairing::arithmetic::FieldExt;
+use std::vec;
+
+use super::param::{KECCAK_WIDTH, PUSH_TABLE_WIDTH};
+
+/// Public data for the bytecode
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct BytecodeRow<F: FieldExt> {
+    hash: F,
+    index: F,
+    is_code: F,
+    byte: F,
+}
+
+/// Unrolled bytecode
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct UnrolledBytecode<F: FieldExt> {
+    bytes: Vec<u8>,
+    rows: Vec<BytecodeRow<F>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Config<F> {
+    r: F,
+    minimum_rows: usize,
+    q_enable: Selector,
+    q_first: Column<Fixed>,
+    q_last: Selector,
+    hash: Column<Advice>,
+    index: Column<Advice>,
+    is_code: Column<Advice>,
+    byte: Column<Advice>,
+    push_rindex: Column<Advice>,
+    hash_rlc: Column<Advice>,
+    hash_length: Column<Advice>,
+    byte_push_size: Column<Advice>,
+    is_final: Column<Advice>,
+    padding: Column<Advice>,
+    push_rindex_inv: Column<Advice>,
+    push_rindex_is_zero: IsZeroConfig<F>,
+    push_table: [Column<Fixed>; PUSH_TABLE_WIDTH],
+    keccak_table: [Column<Advice>; KECCAK_WIDTH],
+}
+
+impl<F: FieldExt> Config<F> {
+    pub(crate) fn configure(meta: &mut ConstraintSystem<F>, r: F) -> Self {
+        let q_enable = meta.complex_selector();
+        let q_first = meta.fixed_column();
+        let q_last = meta.selector();
+        let hash = meta.advice_column();
+        let index = meta.advice_column();
+        let is_code = meta.advice_column();
+        let byte = meta.advice_column();
+        let push_rindex = meta.advice_column();
+        let hash_rlc = meta.advice_column();
+        let hash_length = meta.advice_column();
+        let byte_push_size = meta.advice_column();
+        let is_final = meta.advice_column();
+        let padding = meta.advice_column();
+        let push_rindex_inv = meta.advice_column();
+        let push_table = array_init::array_init(|_| meta.fixed_column());
+        let keccak_table = array_init::array_init(|_| meta.advice_column());
+
+        // A byte is an opcode when `push_rindex == 0` on the previous row,
+        // else it's push data.
+        let push_rindex_is_zero = IsZeroChip::configure(
+            meta,
+            |meta| {
+                // Conditions:
+                // - Not on the first row
+                meta.query_selector(q_enable)
+                    * not::expr(meta.query_fixed(q_first, Rotation::cur()))
+            },
+            |meta| meta.query_advice(push_rindex, Rotation::prev()),
```

Comment:
one question: why not  current rotation `meta.query_advice(push_rindex, Rotation::cur())` when checking `push_rindex`?

Reply:
Certainly possible to do it in different ways, but I think always necessary to check 2 rows for some column because inherently we do not have enough data on a single row (need some history to handle the PUSH data). With how it currently works, the `push_rindex == 0` check is done against the previous row so that `push_rindex` of the current row can store the rindex for the byte in the current row (`push_rindex := is_code ? byte_push_size : push_rindex_prev - 1`). It's also possible I guess to check the value in the current row, and write the value in the next row, which is kind of the same thing but I think complicates things somewhat (because we then need special handling of the last byte which cannot use a value for the next row).

Reply:
sound good to me,  thanks !